json_object.c

```diff
@@ -1214,3 +1214,98 @@ int json_object_equal(struct json_object* jso1, struct json_object* jso2)
 	return 0;
 }
 
+static int json_object_deep_copy_recursive(struct json_object *src, struct json_object **dst)
+{
+	struct json_object *jso = NULL;
+	struct json_object_iter iter;
+	size_t i;
+	int rc = 0;
+
+	if (!src || !dst) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	switch(src->o_type) {
+		case json_type_boolean:
+			*dst = json_object_new_boolean(src->o.c_boolean);
+			break;
+
+		case json_type_double:
+			*dst = json_object_new_double(src->o.c_double);
+			break;
+
+		case json_type_int:
+			*dst = json_object_new_int64(src->o.c_int64);
+			break;
+
+		case json_type_string:
+			*dst = json_object_new_string(get_string_component(src));
+			break;
+
+		case json_type_object:
+			*dst = json_object_new_object();
+			if (!*dst)
+				return -1;
+			json_object_object_foreachC(src, iter) {
+				if ((rc = json_object_deep_copy_recursive(iter.val, &jso)) < 0)
+					return -1;
+				if ((rc = json_object_object_add(*dst, iter.key, jso)) < 0)
+					return -1;
+			}
+			break;
+
+		case json_type_array:
+			*dst = json_object_new_array();
+			if (!*dst)
+				return -1;
+			for (i = 0; i < json_object_array_length(src); i++) {
+				struct json_object *jso1 = json_object_array_get_idx(src, i);
+				if ((rc = json_object_deep_copy_recursive(jso1, &jso)) < 0)
+					return -1;
+				if ((rc = json_object_array_add(*dst, jso)) < 0)
+					return -1;
+			}
+			break;
+
+		case json_type_null:
+			*dst = json_object_new_null();
+			break;
+		default:
+			errno = EINVAL;
+			rc = -1;
+			break;
+	};
+
+	/* errno should be set by the function that faulted */
+	if (!*dst)
+		return -1;
+
+	return rc;
+}
+
+int json_object_deep_copy(struct json_object *src, struct json_object **dst)
+{
+	int rc;
+	/* Nothing to do, so return succes */
+	if (!src && !dst)
+		return 0;
+
+	/* Check if arguments are sane ; refcount must not be higher than 1 */
+	if (!src || !dst || (*dst && (*dst)->_ref_count > 1)) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	/* Free previous destination object */
+	json_object_put(*dst);
```

Comment:
`json_object_put()` handles null;
adding a null-check here may yield a small performance improvement, in the sense that a jump to the `json_object_put()` block, won't happen if that is `null`

most modern compilers would probably optimize this call 
not sure about older compilers & deployment [of json-c on them]

hmm ; i guess i'm neutral about adding or not-adding it;
instead of over-thinking it, i'll just add it